### PR TITLE
Add support for double to varchar coercion in hive tables

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

Allows coerce a double type to varchar in hive tables. 

## Additional context and related issues
Overrides https://github.com/trinodb/trino/pull/18832

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Hive
* Add support for double to varchar coercion in hive tables 
```

### CODE REVIEW

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestDoubleToVarcharCoercions.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        assertThatThrownBy(() -> assertCoercions(DOUBLE, Double.NaN, createVarcharType(1), "NaN", false))
+                .isInstanceOf(TrinoException.class)
+                .hasMessageContaining("Varchar representation of NaN exceeds varchar(1) bounds");
+    }
+
+    @DataProvider
+    public Object[][] doubleValues()
+    {
+        return cartesianProduct(
+                Stream.of(Double.MAX_VALUE, Double.MAX_VALUE, Double.parseDouble("123456789.12345678")).collect(toDataProvider()),
```
Comment (atanasenko) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestDoubleToVarcharCoercions.java:72]: Was there supposed to be MIN_VALUE?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java [hunk 2]
```diff
@@ -395,7 +398,7 @@ protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int pos
         }
     }
 
-    public record CoercionOption(HiveTimestampPrecision timestampPrecision)
+    public record CoercionOption(HiveTimestampPrecision timestampPrecision, boolean treatNaNAsNull)
```
Comment (atanasenko) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:401]: Maybe `CoercionOption` should actually be `CoercionOptions` since it will contain multiple options?

**Review by atanasenko — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestDoubleToVarcharCoercions.java (hunk 1 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestDoubleToVarcharCoercions.java]: Missed it. Added additional value for positive and negative infinity as well.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java (hunk 2 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java]: I have renamed it to `CoercionContext` WDYT ?

**Comment by Praveen2112** [comment]
@atanasenko AC and added some additional tests for infinity.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java [hunk 3]
```diff
@@ -80,12 +80,12 @@ public HivePageSource(
             Optional<BucketValidator> bucketValidator,
             Optional<ReaderProjectionsAdapter> projectionsAdapter,
             TypeManager typeManager,
-            HiveTimestampPrecision timestampPrecision,
+            CoercionContext coercionContext,
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:83]: This belongs to previous commit where you introduced `coercionContext`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSourceProvider.java [hunk 4]
```diff
@@ -213,7 +215,7 @@ public static Optional<ConnectorPageSource> createHivePageSource(
         Optional<BucketAdaptation> bucketAdaptation = createBucketAdaptation(bucketConversion, tableBucketNumber, regularAndInterimColumnMappings);
         Optional<BucketValidator> bucketValidator = createBucketValidator(path, bucketValidation, tableBucketNumber, regularAndInterimColumnMappings);
 
-        CoercionContext coercionContext = new CoercionContext(getTimestampPrecision(session));
+        CoercionContext coercionContext = new CoercionContext(getTimestampPrecision(session), ORC_SERDE_CLASS.equals(getDeserializerClassName(schema)));
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSourceProvider.java:218]: extract variable and say why `treatNaNAsNull` depends on that condition

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java (hunk 3 above)
Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:83]: I have squashed the commit

**Comment by Praveen2112** [comment]
@kokosing AC

**Review by marcinsbd — APPROVED** [review]

**Review by kokosing — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/DoubleToVarcharCoercer.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        double doubleValue = DOUBLE.getDouble(block, position);
+
+        if (Double.isNaN(doubleValue) && treatNaNAsNull) {
+            blockBuilder.appendNull();
+            return;
+        }
+
+        Slice converted = Slices.utf8Slice(Double.toString(doubleValue));
+        if (!toType.isUnbounded() && countCodePoints(converted) > toType.getBoundedLength()) {
+            throw new TrinoException(INVALID_ARGUMENTS, format("Varchar representation of %s exceeds %s bounds", doubleValue, toType));
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/DoubleToVarcharCoercer.java:53]: If we call `truncateToLength(converted, toType)` without calling `countCodePoints(converted) > toType.getBoundedLength()` does that not throw any error ?
It would be nicer to avoid the extra bounds check and have the error come out of `truncateToLength` directly

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestDoubleToVarcharCoercions.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    @DataProvider
+    public Object[][] doubleValues()
+    {
+        return cartesianProduct(
+                Stream.of(
+                        Double.NEGATIVE_INFINITY,
+                        Double.MIN_VALUE,
+                        Double.MAX_VALUE,
+                        Double.POSITIVE_INFINITY,
+                        Double.parseDouble("123456789.12345678")).collect(toDataProvider()),
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestDoubleToVarcharCoercions.java:77]: nit: collect on next line ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSourceProvider.java [hunk 7]
```diff
@@ -191,7 +192,9 @@ public static Optional<ConnectorPageSource> createHivePageSource(
         Optional<BucketAdaptation> bucketAdaptation = createBucketAdaptation(bucketConversion, tableBucketNumber, regularAndInterimColumnMappings);
         Optional<BucketValidator> bucketValidator = createBucketValidator(path, bucketValidation, tableBucketNumber, regularAndInterimColumnMappings);
 
-        CoercionContext coercionContext = new CoercionContext(getTimestampPrecision(session));
+        // ORC files treats Double.NaN as null
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSourceProvider.java:195]: nit: is this an ORC spec behaviour or a Apache Hive behaviour, might want to reword this to clarify

**Comment by Praveen2112** [comment]
Thanks for the review @raunaqmorarka . AC
